### PR TITLE
feat(mentorship): add validation for tags and domains

### LINKED ISSUE #3503: Add Program/Module Tag Validation (Uniqueness & Alphanumeric Security)
## 🎯 Issue Description

This issue addresses a security and data integrity gap in the Mentorship Portal where Program and Module tags currently lack validation. They can be duplicated and may contain potential security risks or inconsistent formatting.
Reference: https://github.com/OWASP/Nest/issues/2701

## 📋 Current Issues

1. **No Uniqueness Validation**: Tags can be duplicated within a single entry (e.g., `["python", "python"]`).
2. **No Format Security**: There is no restriction on tag characters, allowing special characters or potentially unsafe strings.
3. **Data Inconsistency**: Lack of standardization allows for mixed naming conventions (e.g., `["Python", "python"]`).

## 🛠️ Proposed Fix

- Update `CreateProgramInput`, `UpdateProgramInput`, `CreateModuleInput`, and `UpdateModuleInput` to implement validation logic for the `tags` field.
- Enforce uniqueness: Ensure the list of tags contains no duplicates.
- Enforce alphanumeric format: Restrict tags to `a-z`, `A-Z`, and `0-9` using regex validation.
- Add comprehensive unit tests to verify that invalid or duplicate tags are correctly rejected.

**Are you going to work on implementing this?**

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by HarshitVerma109**
## Proposed change

Resolves #3503 

### Validation Logic Implementation
Implemented input validation for tags and domains in the Mentorship app (Programs and Modules) to ensure data consistency and security.

**Core Changes:**
*   Added backend validation `utils.py` to ensure tags and domains are unique and alphanumeric.
*   Integrated this check into Mentorship Program and Module API inputs.
*   Added unit tests to verify valid and invalid inputs.
## Checklist

- [x] **Required:** I followed the [contributing workflow](https://github.com/OWASP/Nest/blob/main/CONTRIBUTING.md#contributing-workflow)
- [x] **Required:** I verified that my code works as intended and resolves the issue as described
- [x] **Required:** I ran `make check-test` locally: all warnings addressed, tests passed
- [ ] I used AI for code, documentation, tests, or communication related to this PR

### CODE REVIEW

**Comment by HarshitVerma109** [comment]
@kasya, please review it and let me know if any changes are required.

File: backend/apps/mentorship/api/internal/nodes/module.py [hunk 1]
```diff
@@ -176,6 +176,15 @@ class CreateModuleInput:
     started_at: datetime
     tags: list[str] = strawberry.field(default_factory=list)
 
+    def __post_init__(self):
+        """Validate input."""
+        from apps.mentorship.api.internal.utils import validate_domains, validate_tags
```
Comment (kasya) [inline backend/apps/mentorship/api/internal/nodes/module.py]: Please move imports to the top of the module

File: backend/apps/mentorship/api/internal/nodes/module.py [hunk 2]
```diff
@@ -194,3 +203,12 @@ class UpdateModuleInput:
     project_name: str
     started_at: datetime
     tags: list[str] = strawberry.field(default_factory=list)
+
+    def __post_init__(self):
+        """Validate input."""
+        from apps.mentorship.api.internal.utils import validate_domains, validate_tags
```
Comment (kasya) [inline backend/apps/mentorship/api/internal/nodes/module.py]: Same here

File: backend/apps/mentorship/api/internal/nodes/program.py [hunk 3]
```diff
@@ -70,6 +70,15 @@ class CreateProgramInput:
     started_at: datetime
     tags: list[str] = strawberry.field(default_factory=list)
 
+    def __post_init__(self):
+        """Validate input."""
+        from apps.mentorship.api.internal.utils import validate_domains, validate_tags
```
Comment (kasya) [inline backend/apps/mentorship/api/internal/nodes/program.py]: Same here

File: backend/apps/mentorship/api/internal/nodes/program.py [hunk 4]
```diff
@@ -86,6 +95,15 @@ class UpdateProgramInput:
     status: ProgramStatusEnum
     tags: list[str] | None = None
 
+    def __post_init__(self):
+        """Validate input."""
+        from apps.mentorship.api.internal.utils import validate_domains, validate_tags
```
Comment (kasya) [inline backend/apps/mentorship/api/internal/nodes/program.py]: same here

**Review by kasya — CHANGES_REQUESTED** [review]
Hi @HarshitVerma109 !
Left some comments ⬇️ 

I'll leave it for @arkid15r to review the rest 👍🏼

**Comment by HarshitVerma109** [comment]
@cubic-dev-ai review this pull request

**Comment by HarshitVerma109** [comment]
@cubic-dev-ai review this pull request

**Comment by HarshitVerma109** [comment]
@arkid15r , please review it and let me know if any changes are required.

**Comment by HarshitVerma109** [comment]
@arkid15r, I’ve resolved the merge conflicts. Please review the changes and let me know if any further adjustments are needed.

File: backend/tests/apps/mentorship/api/internal/nodes/validation_test.py
Comment (arkid15r) [inline backend/tests/apps/mentorship/api/internal/nodes/validation_test.py:1]: The test files structure must mirror application files structure.

File: backend/tests/apps/mentorship/api/internal/nodes/validation_test.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+import pytest
+import strawberry
+
+from apps.mentorship.api.internal.nodes.enum import ExperienceLevelEnum, ProgramStatusEnum
+from apps.mentorship.api.internal.nodes.module import CreateModuleInput, UpdateModuleInput
+from apps.mentorship.api.internal.nodes.program import CreateProgramInput, UpdateProgramInput
+from apps.mentorship.api.internal.utils import validate_domains, validate_tags
+
+
+def test_validate_tags_empty():
```
Comment (arkid15r) [inline backend/tests/apps/mentorship/api/internal/nodes/validation_test.py:14]: We group tests under some class names for better structure.

File: backend/apps/mentorship/api/internal/nodes/program.py [hunk 6]
```diff
@@ -83,6 +84,13 @@ class CreateProgramInput:
     started_at: datetime
     tags: list[str] = strawberry.field(default_factory=list)
 
+    def __post_init__(self):
```
Comment (arkid15r) [inline backend/apps/mentorship/api/internal/nodes/program.py:87]: This implementation is not DRY, please unify program/module domains/tags validation.

**Review by arkid15r — CHANGES_REQUESTED** [review]
